server/src/main/java/com/cloud/server/StatsCollector.java

```diff
@@ -1020,14 +1020,18 @@ protected void runInContext() {
                         if (answer != null && answer.getResult()) {
                             storagePoolStats.put(pool.getId(), (StorageStats)answer);
 
+                            boolean poolNeedsUpdating = false;
                             // Seems like we have dynamically updated the pool size since the prev. size and the current do not match
-                            if (pool.getCapacityBytes() != ((StorageStats)answer).getCapacityBytes() ||
-                                    pool.getUsedBytes() != ((StorageStats)answer).getByteUsed()) {
+                            if (_storagePoolStats.get(poolId) != null && _storagePoolStats.get(poolId).getCapacityBytes() != ((StorageStats)answer).getCapacityBytes()) {
                                 pool.setCapacityBytes(((StorageStats)answer).getCapacityBytes());
-                                if (pool.getStorageProviderName().equalsIgnoreCase(DataStoreProvider.DEFAULT_PRIMARY)) {
-                                    pool.setUsedBytes(((StorageStats) answer).getByteUsed());
-                                    pool.setUpdateTime(new Date());
-                                }
+                                poolNeedsUpdating = true;
+                            }
+                            if (pool.getUsedBytes() != ((StorageStats)answer).getByteUsed() && pool.getStorageProviderName().equalsIgnoreCase(DataStoreProvider.DEFAULT_PRIMARY) && ((StorageStats)answer).getByteUsed() > 0) {
```

Comment:
This restricts updating, when no bytes used in the pool. Suppose, there are few volumes in the pool which are removed later and the answer for 'GetStorageStatsCommand' returns 0 bytes used for that pool.

Reply:
When pool is added to cloudstack or cloudstack is restarted, if the pool is empty the usage will be set to 0. I see that case, let me see what/how can we handle that. This change was recently added in 4.15 only.